pkg/agent/cache/manager.go

```diff
@@ -385,3 +428,103 @@ func (m *manager) getGRPCConn(svid []byte, key *ecdsa.PrivateKey) (*grpc.ClientC
 	}
 	return conn, nil
 }
+
+func (m *manager) rotateBaseSVID() error {
+	m.log.Debug("Checking for BaseSVID expiration")
+	entry := m.getBaseSVIDEntry()
+	if entry.expiry.Sub(time.Now()) < time.Until(entry.expiry)/2 {
+
+		m.log.Debug("Generating new CSR for BaseSVID")
+
+		privateKey, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+		if err != nil {
+			return err
+		}
+
+		csr, err := util.MakeCSR(privateKey, m.baseSPIFFEID)
+		if err != nil {
+			return err
+		}
+
+		conn, err := m.getGRPCConn(entry.svid, entry.key)
+		if err != nil {
+			return err
+		}
+
+		stream, err := node.NewNodeClient(conn).FetchSVID(context.Background())
+		if err != nil {
+			stream.CloseSend()
+			return err
+		}
+
+		m.log.Debug("Sending CSR")
+		err = stream.Send(&node.FetchSVIDRequest{Csrs: [][]byte{csr}})
+		if err != nil {
+			stream.CloseSend()
+			return err
+		}
+		stream.CloseSend()
+
+		resp, err := stream.Recv()
+		if err == io.EOF {
+			return errors.New("FetchSVID stream was empty while trying to rotate BaseSVID")
+		}
+		if err != nil {
+			return err
+		}
+
+		svid := resp.SvidUpdate.Svids[m.baseSPIFFEID]
```

Comment:
Is the base SPIFFE ID guaranteed to be included in the response?

Reply:
It is if there was no error but that could change in the future and it's dangerous assumption to make. I will change it so it handles that possibility.